waspc/src/Wasp/AppSpec/Valid.hs

```diff
@@ -0,0 +1,95 @@
+  case validateExactlyOneAppExists spec of
+    Just err -> [err]
+    Nothing ->
+      -- NOTE: We check these only if App exists because they all rely on it existing.
+      concat
+        [ validateAppAuthIsSetIfAnyPageRequiresAuth spec,
+          validateAuthUserEntityHasCorrectFieldsIfEmailAndPasswordAuthIsUsed spec
+        ]
+
+validateExactlyOneAppExists :: AppSpec -> Maybe ValidationError
+validateExactlyOneAppExists spec =
+  case AS.takeDecls @App (AS.decls spec) of
+    [] -> Just $ GenericValidationError "You are missing an 'app' declaration in your Wasp app."
+    [_] -> Nothing
+    apps ->
+      Just $
+        GenericValidationError $
+          "You have more than one 'app' declaration in your Wasp app. You have " ++ show (length apps) ++ "."
+
+validateAppAuthIsSetIfAnyPageRequiresAuth :: AppSpec -> [ValidationError]
+validateAppAuthIsSetIfAnyPageRequiresAuth spec =
+  if anyPageRequiresAuth && not (isAuthEnabled spec)
+    then
+      [ GenericValidationError
+          "Expected app.auth to be defined since there are Pages with authRequired set to true."
+      ]
+    else []
+  where
+    anyPageRequiresAuth = any ((== Just True) . Page.authRequired) (snd <$> AS.getPages spec)
+
+validateAuthUserEntityHasCorrectFieldsIfEmailAndPasswordAuthIsUsed :: AppSpec -> [ValidationError]
+validateAuthUserEntityHasCorrectFieldsIfEmailAndPasswordAuthIsUsed spec = case App.auth (snd $ getApp spec) of
+  Nothing -> []
+  Just auth ->
+    if Auth.EmailAndPassword `notElem` Auth.methods auth
+      then []
+      else
+        let userEntity = snd $ AS.resolveRef spec (Auth.userEntity auth)
+            userEntityFields = Entity.getFields userEntity
+            maybeEmailField = find ((== "email") . Entity.Field.fieldName) userEntityFields
+            maybePasswordField = find ((== "password") . Entity.Field.fieldName) userEntityFields
+         in concat
+              [ case maybeEmailField of
+                  Just emailField
+                    | Entity.Field.fieldType emailField == Entity.Field.FieldTypeScalar Entity.Field.String -> []
+                  _ ->
+                    [ GenericValidationError
+                        "Expected an Entity referenced by app.auth.userEntity to have field 'email' of type 'string'."
+                    ],
+                case maybePasswordField of
+                  Just passwordField
+                    | Entity.Field.fieldType passwordField == Entity.Field.FieldTypeScalar Entity.Field.String -> []
+                  _ ->
+                    [ GenericValidationError
+                        "Expected an Entity referenced by app.auth.userEntity to have field 'password' of type 'string'."
+                    ]
+              ]
+
+getApp :: AppSpec -> (String, App)
+getApp spec = case takeDecls @App (AS.decls spec) of
+  [app] -> app
+  apps ->
+    error $
+      ("Expected exactly 1 'app' declaration in your wasp code, but you have " ++ show (length apps) ++ ".")
```

Comment:
Good point! I was actually trying to achieve this with "this should never happen" bit, but I agree that it is not strong enough, I like the idea of "internal error" and "bug in Wasp".

I started refactoring this, based on your suggestion, by just modifying this error message here, but then I realized that it would be nice to handle all other `error` invocations in the code. We can safely say that any `error` in Wasp code is really an internal error. Therefore, what I did at the end is catch all of the `error`s at the most external boundary, which is cli/Main.hs. You can see in the code that I catch the `ErrorCall` exceptions, which correspond to `error`, and then I add prefix to them, warning users that this is internal error in Wasp, a bug.

This will now prefix all of the `error` invocations in whole Wasp codebase.